arcgis-ios-sdk-samples/Edit data/Edit features with feature-linked annotation/EditFeaturesWithFeatureLinkedAnnotationViewController.swift

```diff
@@ -0,0 +1,246 @@
+                guard let featureLayer = result.layerContent as? AGSFeatureLayer, let selectedFeature = result.geoElements.first as? AGSFeature else { return }
+                switch selectedFeature.geometry?.geometryType {
+                case .point:
+                    // If the selected feature is a point, prompt the edit attributes alert.
+                    self.showEditableAttributes(selectedFeature: selectedFeature)
+                case .polyline:
+                    let polyline = selectedFeature.geometry as! AGSPolyline
+                    // If the selected feature is a polyline with any part containing more than one segment (i.e. a curve), present an alert.
+                    if polyline.parts.array().contains(where: { $0.pointCount > 2 }) {
+                        self.presentAlert(title: "Make a different selection", message: "Select straight (single segment) polylines only.")
+                        return
+                    }
+                default:
+                    return
+                }
+                // Select the feature.
+                featureLayer.select(selectedFeature)
+                // Set the feature globally.
+                self.selectedFeature = selectedFeature
+            }
+        }
+    }
+    
+    // Create an alert dialog with edit texts to allow editing of the given feature's 'AD_ADDRESS' and 'ST_STR_NAM' attributes.
+    func showEditableAttributes(selectedFeature: AGSFeature) {
+        // Create an alert controller and customize the title and message.
+        let alert = UIAlertController(title: "Edit feature attributes", message: "Edit the 'AD_ADDRESS' and 'ST_STR_NAM' attributes.", preferredStyle: .alert)
+        // Add a "Done" option to complete the editing process and close the alert.
+        let doneAction = UIAlertAction(title: "Done", style: .default) { _ in
+            let addressTextField = alert.textFields?[0]
+            let streetTextField = alert.textFields?[1]
+            // Set the new values for the attributes and update the selected feature.
+            selectedFeature.attributes["AD_ADDRESS"] = Int((addressTextField?.text)!)
+            selectedFeature.attributes["ST_STR_NAM"] = streetTextField?.text
+            selectedFeature.featureTable?.update(selectedFeature)
+        }
+        alert.addAction(doneAction)
+        // Add a text field to prompt the user to input the street address.
+        alert.addTextField { (textField) in
+            // Prompt a number pad for the address.
+            textField.keyboardType = .numberPad
+            // Populate the text fields with the current value.
+            textField.text = (selectedFeature.attributes["AD_ADDRESS"] as! NSNumber).stringValue
+            // Add an observer to ensure the user does not input an empty string.
+            _ = NotificationCenter.default.addObserver(forName: UITextField.textDidChangeNotification, object: textField, queue: OperationQueue.main, using: {_ in
```

Comment:
Should this be removed in deinit?

Reply:
I'm not sure about the deinit, could you advise @philium ?

Reply:
What does the documentation say?

Reply:
The first result from a Google search: https://oleb.net/blog/2018/01/notificationcenter-removeobserver/

Reply:
The documentation says it's not needed for iOS 9.0 and later. Additionally, "If you forget or are unable to remove the observer, the system cleans up the next time it would have posted to it." Seems like it is now personal preference. I see that there are a few samples that do remove the observer. Should I use it in this case?

Reply:
When seeking to understand the behavior of a method, start with the documentation for the method itself:

https://developer.apple.com/documentation/foundation/notificationcenter/1411723-addobserver

You should find your answer there.

Reply:
Yes the Apple doc seems to contradict themselves a bit here. In addObserver it says we must call remove explicitly, while in remove it says it is optional and handled by system... So what I typically did is to call remove to avoid this ambiguity. 🤥

Reply:
There's no contradiction. The documentation for `remove(_:)` says the following:

![image](https://user-images.githubusercontent.com/2257493/90166405-0b4d6d80-dd4f-11ea-816e-82337744f3c2.png)

And the documentation for `addObserver(forName:object:queue:using:)` says this:

![image](https://user-images.githubusercontent.com/2257493/90166575-51a2cc80-dd4f-11ea-98c5-f9e7117e8175.png)

So the automatic unregistration does not apply to this method.

Reply:
Aha! I see - I didn't notice the difference between these 2 `addObserver` signatures. Thanks for clarifying that!